Add ERC: Oracle-Permissioned ERC-20 with ZK Proofs

chadxgpt opened the pull request:
Introduced a new ERC standard that allows token transfers to be validated through an external "Transfer Oracle" using zero-knowledge proofs. This standard aims to enhance compliance and flexibility in institutional tokenization while maintaining ERC-20 compatibility. Included detailed specifications, interfaces, and events for implementation.

chadxeth committed:
Add ERC: Oracle-Permissioned ERC-20 with ZK-Verified ISO 20022 Payment Instructions

Introduced a new ERC standard that allows token transfers to be validated through an external "Transfer Oracle" using zero-knowledge proofs. This standard aims to enhance compliance and flexibility in institutional tokenization while maintaining ERC-20 compatibility. Included detailed specifications, interfaces, and events for implementation.

chadxeth committed:
Enhance ERC-N: Update specifications to use RISC Zero as example instead

chadxeth committed:
Refine ERC-N: Clarify specifications for oracle-permissioned transfers and enhance validation requirements. Added details on approval consumption behavior, amount range design, and best-match selection for overlapping approvals.

chadxeth committed:
Remove proposal attribution.

chadxeth commented on ERCS/erc-N.md:
```diff
@@ -0,0 +1,204 @@
+created: 2025-05-13
+---
+
+Proposed by Ant International: https://www.ant-intl.com/en/
```
Acknowledged and agreed. Removed proposal attribution.

SamWilsn committed:
Update and rename erc-N.md to erc-7963.md

SamWilsn committed:
Rename A&I.jpg to A&I.jpg

SamWilsn commented on ERCS/erc-7963.md:
```diff
@@ -0,0 +1,226 @@
+---
+eip: 7963
+title: Oracle-Permissioned ERC-20 with ZK-Verified ISO 20022 Payment Instructions
```
Unfortunately ISO standards are not generally free ([as in beer](https://en.wikipedia.org/wiki/Gratis_versus_libre#%22Free_beer%22_and_%22freedom_of_speech%22_distinction)) to access. Unless this one is different, we do not permit links (even by name) to non-free resources. Doing so would place a burden on implementers that may be unfair to smaller organizations and individuals who might not be able to access the document.

chadxeth replied:
Hey Sam, thanks for reviewing my PR. I have a couple of questions regarding this that I have compiled from the authors:

1. Is it permissible to mention “ISO 20022 pain.001” by name if the implementation does not require access to the official ISO document? The spec is based on a public interpretation and open-source JSON abstraction.

2. Would a clarification stating that “no ISO standard documents are required to implement this EIP” be sufficient to resolve the concern? I’m happy to include a Licensing & Accessibility section to that effect.

3. Am I allowed to describe the schema as “pain.001-compatible” or “inspired by ISO 20022 pain.001” instead of naming the exact standard?

4. Is referencing the publicly browsable message definitions on iso20022.org acceptable, or should I remove even those?

chadxeth replied:
Hey Sam, do let me know so i can push this forward.

SamWilsn replied:
> Is referencing the publicly browsable message definitions on iso20022.org acceptable, or should I remove even those?

Please remove them.

SamWilsn committed:
Update erc-7963.md

SamWilsn commented on ERCS/erc-7963.md:
```diff
@@ -0,0 +1,226 @@
+*The oracle MAY accept additional public inputs, e.g., extended currency validation, jurisdiction codes, sanctions list epochs*
+
+### Proof System Flexibility
+This standard is **proof-system-agnostic**. The reference implementation uses RISC Zero for:
++ **Transparent Setup**: No trusted ceremony required
++ **Developer Experience**: Write verification logic in Rust
++ **Performance**: Efficient proof generation and verification
++ **Auditability**: Clear, readable verification code
```
Text describing the reference implementation should be placed in that section.

SamWilsn replied:
Still should do this.

SamWilsn commented on ERCS/erc-7963.md:
```diff
@@ -0,0 +1,226 @@
++ **CLI Tools**: Host program for proof generation and verification
++ **Gas Optimization**: Efficient on-chain verification with detailed gas profiling
+
+Repository: [chadxeth/eip-permissioned-erc20](https://github.com/chadxeth/eip-permissioned-erc20)
```
If the reference implementation is small, include it inline in this document. If it is larger, you should include it in the `assets/...` directory under a permissive license.

chadxeth replied:
Understood on this as well. Will include in my next commit.

SamWilsn commented on assets/erc-7963/A&I.jpg:
We prefer SVGs if you can, or PNGs failing that.

SamWilsn replied:
Also this doesn't seem to be used.

chadxeth replied:
I will remove this logo.

chadxeth committed:
Update ERC-7963 title and add description field

- Title: Oracle-Permissioned ERC-20 with ZK Proofs (43 chars)
- Add description field (116 chars, within 140 limit)
- Satisfies EIP-1 format requirements

chadxeth committed:
Add ERC-7963 reference implementation

Complete implementation with:
- Solidity contracts (PermissionedERC20, TransferOracle, RiscZeroVerifier)
- RISC Zero ZK programs for payment instruction validation
- Comprehensive test suite (80 Solidity + 45 Rust tests passing)
- CLI tools and deployment scripts
- Full documentation

Implementation uses generic payment instruction terminology
(refactored from pain.001-specific naming).

Size: 1.0 MB (cleaned of build artifacts and dependencies)

chadxeth committed:
Remove unused image file

chadxeth committed:
Fix EIP validator errors in ERC-7963

- Add required preamble fields: discussions-to, category, requires
- Fix preamble field order (author before status)
- Change section headers from # to ## (Abstract, Motivation, Specification)
- Link EIP-1967 reference
- Remove backticks from file path in Reference Implementation
- Restructure appendix content (remove section, integrate into spec)
- Fix Copyright section with CC0 link
- Change table headers to avoid ERC-7963 pattern matching issues

chadxeth committed:
Fix remaining EIP validator errors

- Move 'requires' field after 'created' in preamble (correct order)
- Reorder sections: Security Considerations now after Reference Implementation
- Simplify reference implementation link to avoid erc-7963 pattern issue

chadxeth commented:
@SamWilsn @bomanaps updates have been made on this EIP taking into account all comments, please do review again thank you.

chadxeth committed:
Fix broken internal links in reference implementation

- Add index.md files for directory navigation in Jekyll build
- Fix LICENSE reference: change from missing file to CC0 statement
- Fix broken EIP-walkthrough.md link: replace with ERC-7963 spec URL
- Update resource links and capitalize Payment Instruction properly

chadxeth committed:
Rename assets directory and fix internal links

- Rename assets/erc-7963/ to assets/eip-7963/ (matches Jekyll convention)
- Update ERC spec to reference ../assets/eip-7963/ paths
- Remove broken directory links from index files (contracts/, test/, etc.)
- Keep only valid links to README.md and zk/ subdirectory

chadxeth commented:
@SamWilsn @bomanaps @fjl gm to all, requesting a review for this, Ant international's team is happy to jump on a call anytime to clarify anything.

SamWilsn reviewed (changes requested):
This pull request is _way_ too large:

<img width="229" height="68" alt="19k lines added" src="https://github.com/user-attachments/assets/16188858-e36b-475c-bf5d-08991701df9d" />

Please reduce the size of your reference implementation to the _minimum_ necessary to illustrate your standard. No build files either please.

chadxeth committed:
Reduce reference implementation to core contracts only

chadxeth committed:
Fix broken internal link to assets directory

chadxeth committed:
Merge branch 'master' into master

chadxeth committed:
Merge branch 'master' into master

chadxeth committed:
Merge branch 'master' into master

chadxgpt committed:
Merge branch 'master' into master

chadxgpt committed:
Merge branch 'master' into master

chadxgpt committed:
Merge branch 'master' into master

chadxgpt committed:
Merge branch 'master' into master

SamWilsn commented on ERCS/erc-7963.md:
```diff
@@ -0,0 +1,274 @@
+
+### ERC-20 Hook
+A _Permissioned ERC-20_ **MUST** replace the standard internal  
+`_update(address from, address to, uint256 amount)` logic with:
```
`_update` is not part of ERC-20, but it is part of [OpenZeppelin](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/5c70e564370f70647ab55b79508904a039a123f1/contracts/token/ERC20/ERC20.sol#L176).

Not everyone uses OpenZeppelin, so you'll want to change how this is defined so that it applies to all ERC-20 implementations, not just ones using OpenZeppelin.

SamWilsn commented on ERCS/erc-7963.md:
````diff
@@ -0,0 +1,274 @@
+```json
+{
+  "messageId": "string",
+  "creationDateTime": "ISO 8601 timestamp",
````
This is an external reference to a source we don't generally permit. I'd recommend, instead, using RFC 3339 or another open date standard.

Alternatively, you can specify how dates are to be formatted without referring to an external standard (but do be careful not to infringe on any IP rights).

SamWilsn commented on ERCS/erc-7963.md:
```diff
@@ -0,0 +1,274 @@
+   - Remove insignificant whitespace
+   - Use minimal JSON encoding
+
+2. **Text Normalization**: Apply UTF-8 NFC (Normalization Form C) to all string fields
```
We allow links to Unicode Technical Standards (see [EIP-1](https://eips.ethereum.org/EIPS/eip-1#unicode-technical-standards-uts)). Please add the correct markdown here for that.

SamWilsn commented on ERCS/erc-7963.md:
```diff
@@ -0,0 +1,274 @@
+3. Maintain the same security guarantees
+
+### Upgradeability
++ Token and Oracle **MAY** be behind [EIP-1967](./eip-1967.md) proxies.
```
It sounds like EIP-1967 might be required reading to understand this proposal, and so should be placed in your `requires:` header.

SamWilsn reviewed (commented):
Looking good! Just a few last comments, and you can get in as draft. Sorry for the long wait!

chadxgpt committed:
Address editor review: open standards, generic hook, requires header

- Reword transfer hook to apply to all ERC-20 implementations, not just
  OpenZeppelin's internal _update
- Replace ISO 8601 date references with RFC 3339 (open standard)
- Link UTF-8 NFC to Unicode Standard Annex #15 (permitted per EIP-1)
- Move RISC Zero reference-implementation description out of the
  Specification into the Reference Implementation section
- Add EIP-1967 to requires header
- Update discussions-to to the Ethereum Magicians thread

chadxgpt committed:
Fix EIP validator: use versioned Unicode Annex #15 link

eipw markdown-rel-links requires the versioned UTS URL form
(tr15/tr15-NN.html); the generic /reports/tr15/ link was rejected.

chadxgpt committed:
Fix author handle: @chadxeth -> @chadxgpt

chadxgpt committed:
Fix EIP validator: reference ERC-1967 with ERC prefix

markdown-refs requires proposals with category ERC (1967) to be
referenced as ERC-1967, not EIP-1967.
